### DISCUSSION TITLE
pl.export_html("...") never finishes

### DISCUSSION THREAD

**Discussion by Galedon1** [pyvista/pyvista#5734] [Q&A]
Hello, 
I am struggling with export_html functionality. Even the basic example provided by the [docs ](https://docs.pyvista.org/version/stable/api/plotting/_autosummary/pyvista.Plotter.export_html.html)never finishes (process active, but CPU at 0% for hours).  I should have trame installed properly, and pyvista updated to current version. Is there a catch in the installation process, or something that I could have missed? 

```
  pv.Report()
  Out[2]: 

  --------------------------------------------------------------------------------
    Date: Wed Mar 06 11:16:25 2024 Střední Evropa (běžný čas)

                OS : Windows
            CPU(s) : 12
           Machine : AMD64
      Architecture : 64bit
               RAM : 31.5 GiB
       Environment : IPython
        GPU Vendor : Intel
      GPU Renderer : Intel(R) UHD Graphics 750
       GPU Version : 4.5.0 - Build 30.0.101.1273
  MathText Support : False

  Python 3.11.8 (tags/v3.11.8:db85d51, Feb  6 2024, 22:03:32) [MSC v.1937 64
  bit (AMD64)]

           pyvista : 0.43.3
               vtk : 9.3.0
             numpy : 1.24.3
        matplotlib : 3.7.3
            scooby : 0.7.2
             pooch : v1.7.0
            pillow : 9.5.0
           imageio : 2.31.6
         pyvistaqt : 0.11.0
             PyQt5 : 5.15.9
           IPython : 8.13.2
        ipywidgets : 8.0.6
             scipy : 1.10.1
              tqdm : 4.65.0
             trame : 3.5.2
      trame_client : 2.15.0
      trame_server : 2.15.0
         trame_vtk : 2.8.5
     trame_vuetify : 2.4.2
      nest_asyncio : 1.5.6

```

**Comment by jourdain** [comment]
The (trame) server needs to be started otherwise the method call will be ignored.

**Reply by Galedon1** [reply]
Thank you for swift answer. But how to do it properly? Simply adding
```
from trame.app import get_server
server = get_server()
server.start()
```
at the beginning of the script doesn't seem to have any effect.
I tried the [Pyvista and Trame](https://tutorial.pyvista.org/tutorial/09_trame/index.html) example, but it want "vue2" instead of "vue3". Adding `server.client_type = "vue2"` only results in `RuntimeError: This event loop is already running`, with additional info: `Output from spyder call 'get_state':
vtk-remote-local-view.size is missing`.

**Reply by jourdain** [reply]
Are you running in jupyter? Can you describe what you are actually doing/running? That might be easier to provide some guidance.

**Reply by jourdain** [reply]
`server.start()` will lock any script, so it is important to understand how you run your python code.

**Reply by Galedon1** [reply]
That would explain a lot - I am using Spyder IDE. Tt uses `%runfile path/to/script.py` in it's console.

**Reply by jourdain** [reply]
Can you share a simple script that I can start from to speed things up on my end? Thx

**Reply by Galedon1** [reply]
That could be those examples I stated before. The example from `Plotter.export_html()` docs with server start added:

```
from trame.app import get_server
server = get_server()
server.start()

import pyvista as pv
from pyvista import examples
mesh = examples.load_uniform()
pl = pv.Plotter(shape=(1, 2))
_ = pl.add_mesh(
    mesh, scalars='Spatial Point Data', show_edges=True
)
pl.subplot(0, 1)
_ = pl.add_mesh(
    mesh, scalars='Spatial Cell Data', show_edges=True
)
pl.export_html('pv.html')  
```
or the trame and pyvista example: 
```
from trame.app import get_server
from trame.ui.vuetify import SinglePageLayout

import pyvista as pv
from pyvista.trame.ui import plotter_ui

# Always set PyVista to plot off screen with Trame
pv.OFF_SCREEN = True

server = get_server()
state, ctrl = server.state, server.controller

mesh = pv.Wavelet()

pl = pv.Plotter()
pl.add_mesh(mesh)

with SinglePageLayout(server) as layout:
    with layout.content:
        # Use PyVista's Trame UI helper method
        #  this will add UI controls
        view = plotter_ui(pl)

server.start()
```

I am running it as a script in Spyder (automatically sending `%runfile path/to/file.py` into the console).

**Reply by jourdain** [reply]
```python
import asyncio
from trame.app import get_server
from trame.ui.vuetify import SinglePageLayout

import pyvista as pv
from pyvista.trame.ui import plotter_ui

# Always set PyVista to plot off screen with Trame
pv.OFF_SCREEN = True

mesh = pv.Wavelet()

pl = pv.Plotter()
pl.add_mesh(mesh)


# PyVista use nest_asyncio for dealing with async
# => pip install nest_asyncio
server = get_server()

@server.controller.add_task("on_server_ready")
async def process(**kwargs):
   pl.export_html("pv.html")
   await server.stop()

server.start()
```

**Reply by Galedon1** [reply]
It raises: 
```
  File d:\path\to\file.py:33
    server.start()

  File ~\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\trame_server\core.py:641 in start
    task = CoreServer.server_start(

  File ~\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\trame_server\protocol.py:50 in server_start
    return server.start_webserver(

  File ~\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\wslink\server.py:301 in start_webserver
    return exec_modes[exec_mode]()

  File ~\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\wslink\server.py:284 in main_exec
    loop.run_until_complete(create_coroutine())

  File C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.2288.0_x64__qbz5n2kfra8p0\Lib\asyncio\base_events.py:630 in run_until_complete
    self._check_running()

  File C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.2288.0_x64__qbz5n2kfra8p0\Lib\asyncio\base_events.py:589 in _check_running
    raise RuntimeError('This event loop is already running')

RuntimeError: This event loop is already running
```
When I run the `server.start()` separately (directely from the console or separating it by `# %%` and using `%runcell -i 1 path/to/file.py`, there is no exception, but also no pv.html (I tried to insert full path into the export_html)

**Reply by jourdain** [reply]
That goes back to my question about how you run it. The example provided assumes it is run as a standalone script. If you are in Jupyter, that is another story as the event loop is already running and you don't need to start the server at all, but you need to properly wait for that (pyvista) server to be ready.

**Reply by Galedon1** [reply]
The above exception raises when I run it as standalone script. I added the other option just because I tried it and thought that it might help. The cell-like running is very similar to the notebook approach, but apparently not in this case.

I use Windows 11,  Python (3.11) installed through Microsoft Store and Spyder (6.0.0.a4) installed using pip.

**Reply by jourdain** [reply]
How come there is an exception on line 33 when the script I shared only has 26 lines? I've downloaded Python 3.11, and the script I shared worked the same.

Here is what I did

```
python3.11 -m venv test-venv
source ./test-venv/bin/activate
pip install "pyvista[all]"
python ./test.py --server
```

And that produced a `pv.html` in the current directory.  

When you run with Spyder, it is running it like in jupyter where an event loop is already running. But the problem is that you can't even use the await keyword at the root like in jupyter, so you can't force some synchronization. Running the server as a task does not solve the problem either. It will require more investigation on how to properly handle async/await within a spyder script.

**Reply by jourdain** [reply]
Actually, it does work when you do the following in spyder, but the file get written somewhere else as the working directory is not the one I was thinking...

```python
async def export():
	server = get_server()
	await DivLayout(server).ready
	pl.export_html("pv9.html")

asyncio.get_event_loop().create_task(export())

```
